File: exercises/full_tank.py
"""
@author Oskar Arensmeier
@date 2024-02-26
"""
import sys

from typing import NamedTuple
from collections import defaultdict

Edge  = NamedTuple("Edge", [("u", int), ("v", int), ("cost", int)])
Query = NamedTuple("Query", [("cap", int), ("start", int), ("end", int)])

def parse_input() -> tuple[int, int, list[int], list[Edge], int, list[Query]]:
    N, M = map(int, sys.stdin.readline().strip().split())

    cities = list(map(int, sys.stdin.readline().strip().split()))
    edges  = []
    for _ in range(M):
        edges.append(Edge(*map(int, sys.stdin.readline().strip().split())))

    q = int(sys.stdin.readline().strip())
    queries = []
    for _ in range(q):
        queries.append(Query(*map(int, sys.stdin.readline().strip().split())))

    return N, M, cities, edges, q, queries

def dfs(
    graph: dict[int, list[int]],
    cities: list[int],
    dists: dict[tuple[int, int], int],
    curr: int,
    end: int,
    cap: int,
    cost: int,
    fuel: int,
    visited: list[bool],
) -> list[int]:
    if curr == end:
        return [cost]

    candidate_solutions = []
    visited[curr] = True
    for f in range(fuel, cap+1):
        for n in graph[curr]:

            if visited[n]:
                continue

            if f < dists[(curr, n)]:
                continue

            candidate_solutions.extend(
                dfs(
                    graph,
                    cities,
                    dists,
                    n,
                    end,
                    cap,
                    cost + cities[curr]*(f - fuel),
                    f - dists[(curr, n)],
                    visited
                )
            )

    visited[curr] = False
    return candidate_solutions

def main() -> int | str:
    N, _, cities, edges, _, queries = parse_input()

    graph = defaultdict(list)
    dists = {}
    for e in edges:
        graph[e.u].append(e.v)
        graph[e.v].append(e.u)
        dists[(e.u, e.v)] = e.cost
        dists[(e.v, e.u)] = e.cost

    solutions = []
    for q in queries:
        candidate_solutions = dfs(
            graph,
            cities,
            dists,
            q.start,
            q.end,
            q.cap,
            0,
            0,
            [False] * (N+1)
        )

        best = float("inf")
        for sol in candidate_solutions:
            best = min(best, sol)

        solutions.append(best)

    return solutions

if __name__ == "__main__":
    for s in main():
        if s == float("inf"):
            print("impossible")
        else:
            print(s)

File: exercises/full_tank.cpp
/**
 * @author Oskar Arensmeier
 * @date 2024-02-27
 */

#include <iostream>
#include <vector>
#include <climits>
#include <bitset>
#include <map>

using namespace std;

struct edge {
    int dest;
    int weight;
};

struct query {
    int start;
    int end;
    int fuel;
};

int N, M, Q;

vector<int> cities;
map<int, vector<edge>> graph;
vector<query> queries;

int dfs(
    int v,
    int fuel,
    int cost,
    int end,
    int cap,
    bitset<1000> & visited
) {
    if (v == end)
        return cost;

    visited[v] = true;
    int best = INT_MAX;
    for (auto & e : graph[v]) {
        if (visited[e.dest])
            continue;

        for (int f = fuel; f <= cap; ++f) {
            int new_cost = cost + cities[v] * (f - fuel);
            int new_fuel = f - e.weight;

            if (new_fuel < 0)
                continue;

            best = min(
                best,
                dfs(
                    e.dest,
                    new_fuel,
                    new_cost,
                    end,
                    cap,
                    visited
                )
            );
        }
    }
    visited[v] = false;

    return best;
}

int main()
{
    cin >> N >> M;

    cities.reserve(N);
    for (int i = 0; i < N; ++i) {
        int c;
        cin >> c;
        cities.push_back(c);
    }

    for (int i = 0; i < M; ++i) {
        int v, u, w;
        cin >> v >> u >> w;

        graph[v].push_back(edge{u, w});
        graph[u].push_back(edge{v, w});
    }

    cin >> Q;
    queries.reserve(Q);
    for (int i = 0; i < Q; ++i) {
        int cap, start, end;
        cin >> cap >> start >> end;

        bitset<1000> visited;
        int res = dfs(start, 0, 0, end, cap, visited);

        if (res == INT_MAX) {
            cout << "impossible\n";
        } else {
            cout << res << '\n';
        }
    }

    cout << flush;
    return 0;
}